File: web/src/test/sessionPace.test.ts
import { describe, expect, it } from 'vitest';

import { buildSessionPace } from '@/domain/workout';

const minute = 60_000;

describe('buildSessionPace', () => {
  it('mide el rango entre la primera y la última serie efectiva', () => {
    const pace = buildSessionPace([
      { createdAtMs: 0, isWarmup: true },
      { createdAtMs: minute, isWarmup: false },
      { createdAtMs: 5 * minute, isWarmup: false },
      { createdAtMs: 11 * minute, isWarmup: false },
    ]);

    expect(pace.firstSetAtMs).toBe(minute);
    expect(pace.lastSetAtMs).toBe(11 * minute);
    expect(pace.spanMs).toBe(10 * minute);
    expect(pace.workingSets).toBe(3);
  });

  it('calcula el ritmo y el descanso promedio', () => {
    const pace = buildSessionPace([
      { createdAtMs: 0, isWarmup: false },
      { createdAtMs: 5 * minute, isWarmup: false },
      { createdAtMs: 10 * minute, isWarmup: false },
    ]);

    expect(pace.setsPerHour).toBeCloseTo(18, 5);
    expect(pace.averageRestMs).toBe(5 * minute);
  });

  it('sin series efectivas no inventa ritmo', () => {
    const pace = buildSessionPace([{ createdAtMs: minute, isWarmup: true }]);

    expect(pace.workingSets).toBe(0);
    expect(pace.setsPerHour).toBeNull();
    expect(pace.averageRestMs).toBeNull();
    expect(pace.firstSetAtMs).toBeNull();
  });

  it('una sola serie no tiene descanso promedio ni ritmo', () => {
    const pace = buildSessionPace([{ createdAtMs: minute, isWarmup: false }]);

    expect(pace.workingSets).toBe(1);
    expect(pace.spanMs).toBe(0);
    expect(pace.averageRestMs).toBeNull();
    expect(pace.setsPerHour).toBeNull();
  });
});

File: android/app/src/test/java/com/fitlog/app/SessionPaceTest.kt
package com.fitlog.app

import com.fitlog.app.domain.PaceSetInput
import com.fitlog.app.domain.WorkoutSummary
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class SessionPaceTest {

    private val minute = 60_000L

    @Test
    fun `mide el rango entre la primera y la ultima serie efectiva`() {
        val pace = WorkoutSummary.pace(
            listOf(
                PaceSetInput(createdAtMs = 0L, isWarmup = true),
                PaceSetInput(createdAtMs = minute, isWarmup = false),
                PaceSetInput(createdAtMs = 5 * minute, isWarmup = false),
                PaceSetInput(createdAtMs = 11 * minute, isWarmup = false),
            )
        )

        assertEquals(minute, pace.firstSetAtMs)
        assertEquals(11 * minute, pace.lastSetAtMs)
        assertEquals(10 * minute, pace.spanMs)
        assertEquals(3, pace.workingSets)
    }

    @Test
    fun `calcula el ritmo y el descanso promedio`() {
        // Tres series efectivas en diez minutos: 18 series por hora y 5 minutos entre series.
        val pace = WorkoutSummary.pace(
            listOf(
                PaceSetInput(createdAtMs = 0L, isWarmup = false),
                PaceSetInput(createdAtMs = 5 * minute, isWarmup = false),
                PaceSetInput(createdAtMs = 10 * minute, isWarmup = false),
            )
        )

        assertEquals(18.0, pace.setsPerHour ?: 0.0, 0.001)
        assertEquals(5 * minute, pace.averageRestMs)
    }

    @Test
    fun `sin series efectivas no inventa ritmo`() {
        val pace = WorkoutSummary.pace(
            listOf(PaceSetInput(createdAtMs = minute, isWarmup = true))
        )

        assertEquals(0, pace.workingSets)
        assertNull(pace.setsPerHour)
        assertNull(pace.averageRestMs)
        assertNull(pace.firstSetAtMs)
    }

    @Test
    fun `una sola serie no tiene descanso promedio ni ritmo`() {
        val pace = WorkoutSummary.pace(
            listOf(PaceSetInput(createdAtMs = minute, isWarmup = false))
        )

        assertEquals(1, pace.workingSets)
        assertEquals(0L, pace.spanMs)
        assertNull(pace.averageRestMs)
        assertNull(pace.setsPerHour)
    }
}
